cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactDescriptor.java

```diff
@@ -19,17 +19,27 @@
 import io.cdap.cdap.api.artifact.ArtifactId;
 import org.apache.twill.filesystem.Location;
 
+import java.net.URI;
+
 /**
  * Uniquely describes an artifact. Artifact descriptors are ordered by scope,
  * then by name, and finally by version.
  */
 public final class ArtifactDescriptor implements Comparable<ArtifactDescriptor> {
   private final ArtifactId artifactId;
-  private final Location location;
+
+  /**
+   * Mark with transient to not serialize it. Store the location URI in separate field "locationURI" that
+   * gets serialized when this object needs to be transferred through the network.
+   * The recipient should instantiate a {@link Location} based on the received "locationURI"
+   */
+  private final transient Location location;
```

Comment:
why do we make this change? Is this class used anywhere in the response? If so, this is changing the API behavior

Reply:
This PR introduces a REST endpoint for getting artifact detail, thus requiring support for serialization and deserialization. But Location is an interface that cannot be sterilized easily, thus ignoring this field and using location URI instead. The requester would need to convert from Location URI to Location using proper location factory.

Reply:
This means when it get deserialized, the location will be null, which makes this class unreliable, is it necessary to return the entire ArtifactDetail back in the REST call, can we just return a subset of them?

Reply:
This means when it get deserialized, the location will be null, which makes this class unreliable, is it necessary to return the entire ArtifactDetail back in the REST call, can we just return a subset of them?

Reply:
Correct on location will be null. But the Remote repo reader client side implication inits Location so it returns a complete ArtifactDetail to the caller. 

Actually Location object is one of the key info we need to fetch from remote for preview.

Reply:
Better to put a comment about this, otherwise it will be super confusing to future development

Reply:
Okay. Added some more to the description.